### ISSUE TITLE
Support Sqlite storage type

### ISSUE THREAD

**Issue by Half-Shot** [matrix-org/matrix-rust-sdk-crypto-wasm#168]
As part of my quest to look at moving the bots and bridges away from the nodejs binding layer, I'd like to talk about potentially enabling SQLite support for this binding layer. In the NodeJS project, we do https://github.com/matrix-org/matrix-rust-sdk-crypto-nodejs/blob/main/src/machine.rs#L113-L155.

**Comment by richvdh** [comment]
I guess this would require the use of [Sqlite's support for wasm](https://sqlite.org/wasm/doc/trunk/index.md). Not sure what would be involved in wiring that up.

**Comment by Hywan** [comment]
I don't think we should use Sqlite compiled as Wasm. Instead, we should expose SQLite API via Wasm imports, and use SQLite compiled as a NodeJS bindings from C. It would be probably much faster and easier.

**Comment by richvdh** [comment]
> Instead, we should expose SQLite API via Wasm imports

You mean, build our own wrapper around the sqlite API, and somehow thread that through to the rust layer? That sounds like a lot of work.

**Comment by Hywan** [comment]
It's probably a lot of work because the Matrix Rust SDK expects rusqlite API entirely. I wonder how it compiles to `wasm32-unknown-unknown` and how linking is done. We should try.

It's probably more work than maintaining `matrix-rust-sdk-crypto-nodejs` if you ask me 😛.

**Comment by Half-Shot** [comment]
Well, the issue exists because nobody really maintains `matrix-rust-sdk-crypto-nodejs` right now and that leads to the integrations lagging behind the clients, causing people to pull out their hair.

But if doing the work to make our integrations support the maintained wasm bindings is **more** work than just having someone maintain the nodejs bindings, then that's fine too. The current situation is untenable though.

**Comment by richvdh** [comment]
Other than the overhead of maintaining two sets of bindings, another advantage of supporting sqlite through the wasm bindings rather than maintaining matrix-rust-sdk-crypto-nodejs would be that you could also use matrix-js-sdk for nodejs projects.

**Comment by richvdh** [comment]
Yet another option here might be to use https://www.npmjs.com/package/indexeddb, which implements indexeddb backed by the local filesystem. That said, its 35 weekly downloads aren't entirely reassuring.

**Comment by Half-Shot** [comment]
![image](https://github.com/user-attachments/assets/d1668a51-3dae-4ef1-a7ab-1d5c254d21af)

it's vintage :wine_glass:

**Comment by richvdh** [comment]
In fairness 0.0.1 was *only* published 4 years ago. I guess if it works, don't change it?

**Comment by Half-Shot** [comment]
One option is https://github.com/tursodatabase/turso for the sqlite-ing. It claims to compile nicely to WASM and is SQLite compatible, but equally it's [**not** SQLite](https://github.com/tursodatabase/turso/blob/main/COMPAT.md) and we'd be A) feature limited B) using beta software.

Whether this is worse than using a indexddb impl, or https://docs.rs/sqlite-wasm-rs/latest/sqlite_wasm_rs/, I am not sure.

But I do think killing the nodejs binding layer is important at this stage.
